presto-main/src/main/java/com/facebook/presto/metadata/SqlScalarFunction.java

```diff
@@ -112,13 +112,29 @@ protected SqlScalarFunction(
                 literalParameters);
     }
 
+    protected SqlScalarFunction(Signature signature)
+    {
+        this.signature = requireNonNull(signature, "signature is null");
+        checkArgument(signature.getKind() == SCALAR, "function kind must be SCALAR");
+    }
+
     @Override
     public final Signature getSignature()
     {
         return signature;
     }
 
-    public abstract ScalarFunctionImplementation specialize(Map<String, Type> types, int arity, TypeManager typeManager, FunctionRegistry functionRegistry);
+    public abstract ScalarFunctionImplementation specialize(Map<String, Type> types, List<TypeSignature> parameterTypes, TypeManager typeManager, FunctionRegistry functionRegistry);
```

Comment:
We are working on implicit coercions for Decimal type in a separate branch (https://github.com/Teradata/presto/pull/119). As part of that work we did a lot of refactorings. Among them we changed the `specialize` signature. Now it accepts `BoundVariables` class instead of arguments, that contains both "types" and "literal" variables bindings. Also during the migration of `Signature` tests we refactored assertion in there, what resolves (https://github.com/facebook/presto/issues/4405). We can merge decimal pull request with the decimal coercions one if you wish. So this particular issue will be resolved.